File: experiments-local/experiment-19/runs/agent=omp_language=csharp_model=claude-sonnet-4-6_tooling=none/rep2/BrazilianSoccerMcp/Tools/StatsTools.cs
using System.ComponentModel;
using System.Text;
using BrazilianSoccerMcp.Models;
using BrazilianSoccerMcp.Services;
using ModelContextProtocol.Server;

namespace BrazilianSoccerMcp.Tools;

[McpServerToolType]
public sealed class StatsTools(DataRepository repo)
{
    [McpServerTool(Name = "get_team_statistics")]
    [Description(
        "Get win/draw/loss record, goals scored/conceded, and points for a team. " +
        "Optionally filter by season and competition.")]
    public string GetTeamStatistics(
        [Description("Team name (partial match). E.g. 'Flamengo', 'Corinthians'")] string team,
        [Description("Season year. E.g. 2023")] int? season = null,
        [Description("Competition: Brasileirao, CopaDoBrasil, Libertadores, BrFootball, HistoricoBrasileiro")] string? competition = null)
    {
        var comp = ParseCompetition(competition);
        var stats = repo.GetTeamStats(team, season, comp);

        if (stats.Matches == 0)
            return $"No matches found for '{team}'" +
                   (season.HasValue ? $" in {season}" : "") +
                   (competition != null ? $" ({competition})" : "") + ".";

        var sb = new StringBuilder();
        sb.AppendLine($"Statistics for {stats.Team}" +
            (season.HasValue ? $" — {season}" : "") +
            (competition != null ? $" — {competition}" : ""));
        sb.AppendLine();
        sb.AppendLine($"  Matches:  {stats.Matches}");
        sb.AppendLine($"  Record:   {stats.Wins}W / {stats.Draws}D / {stats.Losses}L");
        sb.AppendLine($"  Points:   {stats.Points}");
        sb.AppendLine($"  Goals:    {stats.GoalsFor} scored, {stats.GoalsAgainst} conceded (GD: {stats.GoalDifference:+#;-#;0})");
        sb.AppendLine($"  Win rate: {stats.WinRate:F1}%");

        return sb.ToString();
    }

    [McpServerTool(Name = "get_standings")]
    [Description(
        "Calculate league standings for a season from match results. " +
        "Returns ranked table with points, W/D/L, goals.")]
    public string GetStandings(
        [Description("Season year, e.g. 2023")] int season,
        [Description("Competition (default: Brasileirao). Options: Brasileirao, HistoricoBrasileiro")] string competition = "Brasileirao")
    {
        var comp = ParseCompetition(competition) ?? Competition.Brasileirao;
        var standings = repo.GetStandings(season, comp);

        if (standings.Count == 0)
            return $"No match data found for {season} {competition}.";

        var sb = new StringBuilder();
        sb.AppendLine($"{season} {comp} Standings (calculated from match results):");
        sb.AppendLine();
        sb.AppendLine($"  {"Pos",-4} {"Team",-30} {"Pts",-5} {"MP",-4} {"W",-4} {"D",-4} {"L",-4} {"GF",-4} {"GA",-4} {"GD",-5}");
        sb.AppendLine(new string('-', 75));

        int pos = 1;
        foreach (var (stats, teamName) in standings)
        {
            var gd = stats.GoalDifference >= 0 ? $"+{stats.GoalDifference}" : $"{stats.GoalDifference}";
            sb.AppendLine(
                $"  {pos,-4} {teamName,-30} {stats.Points,-5} {stats.Matches,-4} {stats.Wins,-4} {stats.Draws,-4} {stats.Losses,-4} {stats.GoalsFor,-4} {stats.GoalsAgainst,-4} {gd,-5}");
            pos++;
        }

        return sb.ToString();
    }

    [McpServerTool(Name = "get_biggest_wins")]
    [Description(
        "Get the biggest victories (by goal difference) in the dataset. " +
        "Optionally filter by competition.")]
    public string GetBiggestWins(
        [Description("Competition filter: Brasileirao, CopaDoBrasil, Libertadores, BrFootball, HistoricoBrasileiro")] string? competition = null,
        [Description("Number of results to return (default 10)")] int limit = 10)
    {
        var comp = ParseCompetition(competition);
        var matches = repo.GetBiggestWins(comp, limit);

        if (matches.Count == 0)
            return "No matches found.";

        var sb = new StringBuilder();
        sb.AppendLine($"Biggest victories{(competition != null ? $" in {competition}" : "")} (by goal difference):");
        sb.AppendLine();

        int rank = 1;
        foreach (var m in matches)
        {
            var winner = m.HomeGoals > m.AwayGoals
                ? TeamNameNormalizer.Normalize(m.HomeTeam)
                : TeamNameNormalizer.Normalize(m.AwayTeam);
            sb.AppendLine(
                $"  {rank++,2}. {m.Date:yyyy-MM-dd}  {TeamNameNormalizer.Normalize(m.HomeTeam)} {m.HomeGoals}-{m.AwayGoals} {TeamNameNormalizer.Normalize(m.AwayTeam)}");
            sb.AppendLine($"      Winner by {m.GoalDifference} | {m.CompetitionLabel} {m.Season}" +
                (m.Stage != null ? $" | {m.Stage}" : ""));
        }

        return sb.ToString();
    }

    [McpServerTool(Name = "get_competition_stats")]
    [Description(
        "Get aggregate statistics for a competition: average goals per match, " +
        "home/away/draw rates, total matches.")]
    public string GetCompetitionStats(
        [Description("Competition: Brasileirao, CopaDoBrasil, Libertadores, BrFootball, HistoricoBrasileiro. Leave null for all.")] string? competition = null,
        [Description("Season year filter. E.g. 2023")] int? season = null)
    {
        var comp = ParseCompetition(competition);
        var (avgGoals, homeWinRate, drawRate, awayWinRate, total) =
            repo.GetCompetitionStats(comp, season);

        if (total == 0)
            return "No data found for the given criteria.";

        var sb = new StringBuilder();
        sb.AppendLine($"Competition statistics:" +
            (competition != null ? $" {competition}" : " (all competitions)") +
            (season.HasValue ? $" — {season}" : ""));
        sb.AppendLine();
        sb.AppendLine($"  Total matches:       {total}");
        sb.AppendLine($"  Avg goals per match: {avgGoals:F2}");
        sb.AppendLine($"  Home win rate:       {homeWinRate:F1}%");
        sb.AppendLine($"  Draw rate:           {drawRate:F1}%");
        sb.AppendLine($"  Away win rate:       {awayWinRate:F1}%");

        return sb.ToString();
    }

    private static Competition? ParseCompetition(string? s) => s?.ToLowerInvariant().Replace(" ", "") switch
    {
        "brasileirao" or "brasileirão" or "seriea" => Competition.Brasileirao,
        "copodobrasil" or "copa" or "cup" => Competition.CopaDoBrasil,
        "libertadores" or "copalibertadores" => Competition.Libertadores,
        "brfootball" => Competition.BrFootball,
        "historico" or "historicobrasileiro" or "historicobrasileirao" => Competition.HistoricoBrasileiro,
        null => null,
        _ => null,
    };
}

File: experiments-local/experiment-mu-sonnet-claudecode/runs/agent=claude-code_language=typescript_model=sonnet-4.6_tooling=none/rep3/src/tools/stats-tools.ts
import { getDataStore, teamMatches } from "../data-loader.js";

export function getAggregateStats(params: { competition?: string; season?: number }): string {
  const { matches } = getDataStore();

  let filtered = matches.filter((m) => {
    if (params.competition) {
      const comp = params.competition.toLowerCase();
      if ((comp.includes("brasileirao") || comp.includes("serie a")) && m.competition !== "brasileirao" && m.competition !== "historical") return false;
      if (comp.includes("copa do brasil") && m.competition !== "copa_do_brasil") return false;
      if (comp.includes("libertadores") && m.competition !== "libertadores") return false;
    }
    if (params.season && m.season !== params.season) return false;
    return true;
  });

  if (filtered.length === 0) return "No matches found for the given criteria.";

  let homeWins = 0, awayWins = 0, draws = 0;
  let totalGoals = 0;
  let highestScoring = filtered[0];
  let lowestScoring = filtered[0];

  for (const m of filtered) {
    totalGoals += m.homeGoals + m.awayGoals;
    const total = m.homeGoals + m.awayGoals;
    if (total > (highestScoring.homeGoals + highestScoring.awayGoals)) highestScoring = m;
    if (total < (lowestScoring.homeGoals + lowestScoring.awayGoals)) lowestScoring = m;
    if (m.homeGoals > m.awayGoals) homeWins++;
    else if (m.homeGoals < m.awayGoals) awayWins++;
    else draws++;
  }

  const total = filtered.length;
  const avgGoals = (totalGoals / total).toFixed(2);
  const homeWinRate = ((homeWins / total) * 100).toFixed(1);
  const awayWinRate = ((awayWins / total) * 100).toFixed(1);
  const drawRate = ((draws / total) * 100).toFixed(1);

  const context = [
    params.competition ?? "All competitions",
    params.season ? `Season ${params.season}` : "All seasons",
  ].join(", ");

  return [
    `Aggregate Statistics (${context})`,
    `Total Matches: ${total}`,
    `Total Goals: ${totalGoals}`,
    `Average Goals/Match: ${avgGoals}`,
    `Home Wins: ${homeWins} (${homeWinRate}%)`,
    `Away Wins: ${awayWins} (${awayWinRate}%)`,
    `Draws: ${draws} (${drawRate}%)`,
    `Highest Scoring Match: ${highestScoring.homeTeam} ${highestScoring.homeGoals}-${highestScoring.awayGoals} ${highestScoring.awayTeam} (${highestScoring.homeGoals + highestScoring.awayGoals} goals)`,
  ].join("\n");
}

export function getSeasonComparison(params: { season1: number; season2: number; competition?: string }): string {
  const { matches } = getDataStore();

  const getSeasonStats = (season: number) => {
    const filtered = matches.filter((m) => {
      if (m.season !== season) return false;
      if (params.competition) {
        const comp = params.competition.toLowerCase();
        if ((comp.includes("brasileirao") || comp.includes("serie a")) && m.competition !== "brasileirao" && m.competition !== "historical") return false;
      } else {
        if (m.competition !== "brasileirao" && m.competition !== "historical") return false;
      }
      return true;
    });

    const goals = filtered.reduce((s, m) => s + m.homeGoals + m.awayGoals, 0);
    const homeWins = filtered.filter((m) => m.homeGoals > m.awayGoals).length;
    const teams = new Set([...filtered.map((m) => m.homeTeam), ...filtered.map((m) => m.awayTeam)]);

    return {
      season,
      matches: filtered.length,
      goals,
      avgGoals: filtered.length > 0 ? (goals / filtered.length).toFixed(2) : "0",
      homeWinRate: filtered.length > 0 ? ((homeWins / filtered.length) * 100).toFixed(1) : "0",
      teams: teams.size,
    };
  };

  const s1 = getSeasonStats(params.season1);
  const s2 = getSeasonStats(params.season2);

  if (s1.matches === 0 && s2.matches === 0) {
    return `No data found for seasons ${params.season1} or ${params.season2}.`;
  }

  const comp = params.competition ?? "Brasileirão";

  const formatRow = (label: string, v1: string | number, v2: string | number) =>
    `${label.padEnd(22)} ${String(v1).padStart(10)}   ${String(v2).padStart(10)}`;

  return [
    `${comp} Season Comparison`,
    formatRow("", params.season1, params.season2),
    "-".repeat(47),
    formatRow("Matches", s1.matches, s2.matches),
    formatRow("Total Goals", s1.goals, s2.goals),
    formatRow("Avg Goals/Match", s1.avgGoals, s2.avgGoals),
    formatRow("Home Win Rate", `${s1.homeWinRate}%`, `${s2.homeWinRate}%`),
    formatRow("Teams", s1.teams, s2.teams),
  ].join("\n");
}

export function getMostGoals(params: { competition?: string; season?: number; limit?: number }): string {
  const { matches } = getDataStore();
  const limit = params.limit ?? 10;

  let filtered = matches.filter((m) => {
    if (params.competition) {
      const comp = params.competition.toLowerCase();
      if ((comp.includes("brasileirao") || comp.includes("serie a")) && m.competition !== "brasileirao" && m.competition !== "historical") return false;
      if (comp.includes("copa do brasil") && m.competition !== "copa_do_brasil") return false;
      if (comp.includes("libertadores") && m.competition !== "libertadores") return false;
    }
    if (params.season && m.season !== params.season) return false;
    return true;
  });

  // Goals scored by team
  const teamGoals = new Map<string, number>();
  for (const m of filtered) {
    const hk = m.homeTeam.toLowerCase();
    const ak = m.awayTeam.toLowerCase();
    teamGoals.set(hk, (teamGoals.get(hk) ?? 0) + m.homeGoals);
    teamGoals.set(ak, (teamGoals.get(ak) ?? 0) + m.awayGoals);
  }

  // Get display name for each team key
  const teamNames = new Map<string, string>();
  for (const m of filtered) {
    teamNames.set(m.homeTeam.toLowerCase(), m.homeTeam);
    teamNames.set(m.awayTeam.toLowerCase(), m.awayTeam);
  }

  const sorted = Array.from(teamGoals.entries())
    .map(([key, goals]) => ({ team: teamNames.get(key) ?? key, goals }))
    .sort((a, b) => b.goals - a.goals)
    .slice(0, limit);

  const context = [
    params.competition ?? "All competitions",
    params.season ? `Season ${params.season}` : "All seasons",
  ].join(", ");

  const lines = sorted.map((t, i) => `${i + 1}. ${t.team}: ${t.goals} goals`);
  return [`Top Goal-Scoring Teams (${context})`, ...lines].join("\n");
}
